### PR TITLE
Adding a configurable directory for saving compiled model artifacts

### PR DESCRIPTION
**Pull request by dhawalkp**
Issue: https://github.com/facebookincubator/AITemplate/issues/396
StableDiffusionAITPipeline and StableDiffusionImg2ImgAITPipeline in pipeline_stable_diffusion_ait.py and [pipeline_stable_diffusion_img2img_ait.py has workdir hardcoded to "tmp/" which might create problems where the file system locations are either read-only or users have restricted permissions in the current working directory. Examples include managed ML services in Cloud that might have file permission restrictions.

Desired behavior: User should be given flexibility to specify the custom location as workdir where AITemplate .so files are downloaded.

### CODE REVIEW

**Review by lanking520 — APPROVED** [review]
LGTM. This approach solved a lot of custom hack on the path issue after model conversion

File: examples/05_stable_diffusion/scripts/compile.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/compile.py:62]: I've noticed that this code is duplicated across a bunch of files. Can it be deduped and put into a helper function instead?

**Review by khabinov — CHANGES_REQUESTED** [review]
Thanks @dhawalkp! The change looks ok, however I left one comment about code duplication.

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py:62]: I thought about it however, I did not pursue to decouple it into a seperate function is due to the fact, some code expects to load the models from specific folder path like "diffusers-pipeline/stabilityai.." vs other code expects just the parent directory with "tmp/". Thus, in order to keep the existing code intact and retain backward compatibility, I chose this approach.

Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py:62]: We can work on changing this and decoupling it as a second iteration on top of this PR. Right now, as it was a burning issue for us and was a blocker, had to push this PR.

File: examples/05_stable_diffusion/scripts/demo_img2img.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/demo_img2img.py:61]: Ditto about 1) indentation, 2) usage of try/except, 3) spaces between args, 4) string formatting.

File: examples/05_stable_diffusion/scripts/demo_img2img.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/demo_img2img.py:47]: Nit, maybe avoid the extra lines.

File: examples/05_stable_diffusion/scripts/demo_img2img.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/demo_img2img.py:16]: Ditto about alphabetical order.

File: examples/05_stable_diffusion/scripts/download_pipeline.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/download_pipeline.py:17]: Ditto about alphabetical order.

File: examples/05_stable_diffusion/src/compile_lib/compile_clip.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/compile_lib/compile_clip.py:17]: Ditto about alphabetical order.

File: examples/05_stable_diffusion/scripts/download_pipeline.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/download_pipeline.py:44]: Ditto about 1) indentation, 2) usage of try/except, 3) spaces between args, 4) string formatting.

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:74]: Ditto about 1) indentation, 2) usage of try/except, 3) spaces between args, 4) string formatting.

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:160]: Ditto about 1) usage of try/except, 2) string formatting.

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:242]: Ditto about 1) usage of try/except, 2) string formatting.

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:325]: Ditto about 1) usage of try/except, 2) string formatting.

File: examples/05_stable_diffusion/src/compile_lib/compile_clip.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/compile_lib/compile_clip.py:137]: Ditto about 1) indentation, 2) usage of try/except, 3) spaces between args, 4) string formatting.

File: examples/05_stable_diffusion/src/compile_lib/compile_unet.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/compile_lib/compile_unet.py:104]: Ditto about 1) indentation, 2) usage of try/except, 3) spaces between args, 4) string formatting.

File: examples/05_stable_diffusion/src/compile_lib/compile_vae.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/compile_lib/compile_vae.py:151]: Ditto about 1) indentation, 2) usage of try/except, 3) spaces between args, 4) string formatting.

File: examples/05_stable_diffusion/src/pipeline_stable_diffusion_ait.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/pipeline_stable_diffusion_ait.py:114]: Ditto about 1) usage of try/except, 2) string formatting.

File: examples/05_stable_diffusion/src/pipeline_stable_diffusion_img2img_ait.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/pipeline_stable_diffusion_img2img_ait.py:127]: Ditto about 1) usage of try/except, 2) string formatting.

**Review by khabinov — CHANGES_REQUESTED** [review]
There is a bunch of nits, it would be nice if they are addressed. Thanks for understanding!

File: examples/05_stable_diffusion/src/pipeline_stable_diffusion_img2img_ait.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/pipeline_stable_diffusion_img2img_ait.py]: done

File: examples/05_stable_diffusion/src/pipeline_stable_diffusion_ait.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/pipeline_stable_diffusion_ait.py]: Done

File: examples/05_stable_diffusion/src/compile_lib/compile_vae.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/compile_lib/compile_vae.py]: Done

File: examples/05_stable_diffusion/src/compile_lib/compile_unet.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/compile_lib/compile_unet.py]: Done

File: examples/05_stable_diffusion/src/compile_lib/compile_clip.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/compile_lib/compile_clip.py]: Done

File: examples/05_stable_diffusion/scripts/demo.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/demo.py:18]: Done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: Done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: Done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: Done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: Done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/compile.py:18]: Is `os` module still needed here?

File: examples/05_stable_diffusion/scripts/compile.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/compile.py:52]: `which has AITemplate compiled artifacts the model(s).` feels like some word is missing here.

File: examples/05_stable_diffusion/scripts/compile.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/compile.py:54]: `By default, the it will look` is "the" word needed here?

File: examples/05_stable_diffusion/scripts/demo.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/demo.py:18]: Same as above

File: examples/05_stable_diffusion/scripts/compile.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/compile.py:56]: Nit, maybe this line can be removed?

File: examples/05_stable_diffusion/scripts/demo.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/demo.py:45]: Nit, these lines can be removed

File: examples/05_stable_diffusion/scripts/demo.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/demo.py:52]: This line can be removed

File: examples/05_stable_diffusion/scripts/download_pipeline.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/download_pipeline.py:20]: Nit, can this be moved into the `__main__`?

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:74]: Nit, maybe we can avoid having capital letters in function names?

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:160]: Same as above about capital letters.

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:245]: Same as above about capital letters.

File: examples/05_stable_diffusion/src/benchmark.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/benchmark.py:331]: Wondering whether it is necessary to have this repeated comment? Can it simply be moved into the source code of `get_work_dir_location_diffusers()`?

File: examples/05_stable_diffusion/src/compile_lib/compile_clip.py
Comment (khabinov) [inline examples/05_stable_diffusion/src/compile_lib/compile_clip.py:22]: Why empty line?

**Review by khabinov — CHANGES_REQUESTED** [review]
Thanks @dhawalkp! I think it looks much better now, just a few more suggestions to address.

File: examples/05_stable_diffusion/src/compile_lib/compile_clip.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/compile_lib/compile_clip.py:21]: Done

File: examples/05_stable_diffusion/src/benchmark.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/benchmark.py]: Done

File: examples/05_stable_diffusion/src/benchmark.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/benchmark.py]: Done

File: examples/05_stable_diffusion/src/benchmark.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/benchmark.py]: done

File: examples/05_stable_diffusion/src/benchmark.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/src/benchmark.py]: done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: done

File: examples/05_stable_diffusion/scripts/demo.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/demo.py]: done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: done

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py]: done

**Comment by dhawalkp** [comment]
@khabinov Is it good to merge?

**Comment by dhawalkp** [comment]
hi @khabinov can you please approve the pull request. Appreciate if we can get this merged today so that we can unblock our customers.

File: examples/05_stable_diffusion/scripts/download_pipeline.py
Comment (khabinov) [reply examples/05_stable_diffusion/scripts/download_pipeline.py:19]: What is blocking this from being moved into `if __name__ == "__main__"` section?

File: examples/05_stable_diffusion/scripts/compile.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/compile.py:49]: There is also an internal question about `--local-dir` option:

> Do we need to consider removing the option similar to here? What if user set it here but get override in get_work_dir_location_diffusers() ?

cc @frank-wei

**Review by khabinov — CHANGES_REQUESTED** [review]
Hi @dhawalkp,

> Appreciate if we can get this merged today so that we can unblock our customers.

Could you please take a look at a couple of comments? Is it possible for you to use code from your branch until this PR is merged? Our internal CI takes quite some time.

**Comment by dhawalkp** [comment]
> 

Fixed this CI comment.

File: examples/05_stable_diffusion/scripts/compile.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/compile.py:44]: Addressed both the comments.

File: examples/05_stable_diffusion/scripts/download_pipeline.py
Comment (khabinov) [inline examples/05_stable_diffusion/scripts/download_pipeline.py:27]: Seems like the linter doesn't like the empty line

**Review by khabinov — CHANGES_REQUESTED** [review]

File: examples/05_stable_diffusion/scripts/download_pipeline.py
Comment (dhawalkp) [reply examples/05_stable_diffusion/scripts/download_pipeline.py]: Done

**Comment by dhawalkp** [comment]
@khabinov I executed the test_correctness.py but not seeing any failure. Could you point me to the logs in your test that can give me more info as to why the test failed?

**Comment by dhawalkp** [comment]
@khabinov any suggestions?

**Comment by khabinov** [comment]
@terrychenism started to look into this, he should be able to help.

**Comment by dhawalkp** [comment]
@khabinov Thank you so much! Looking forward to resolving this and merging my PR again

**Comment by dhawalkp** [comment]
@terrychenism @khabinov  any updates on this? How can I change the pull request to incorporate this feedback?

**Comment by dhawalkp** [comment]
@khabinov @terrychenism any updates?

**Comment by terrychenism** [comment]
Could you refactor this PR a bit more, e.g create class for UNET/CLIP/VAE, then load the so from your assigned dir, without loading pytorch weight.

**Comment by dhawalkp** [comment]
@terrychenism not sure I understand fully. Are you referring to utils.py changes? If you can show me sample on what exactly are you asking would allow me to understand it clearly. Thanks

**Comment by dhawalkp** [comment]
Also, which test case failed internally that @khabinov was referring to?

**Pull request closed without merging**
